src/csharp/Microsoft.Spark/Utils/PythonSerDe.cs

```diff
@@ -67,7 +68,48 @@ internal static object[] GetUnpickledObjects(Stream stream, int messageLength)
                     stackCapacity: 102); // Spark sends batches of 100 rows, and +2 is for markers.
                 s_rowConstructor.Reset();
                 Debug.Assert(unpickledItems != null);
-                return (unpickledItems as object[]);
+                // Check if unpickler returns ArrayList.
+                // If so, it needs to be cast to the appropriate array type.
+                foreach (object objArr in (object[])unpickledItems)
+                {
+                    if (objArr.GetType() == typeof(object[]))
+                    {
+                        object obj = ((object[])objArr)[0];
+                        // Check null case.
+                        if (obj == null)
+                        {
+                            continue;
+                        }
+                        // Check if obj is ArrayList, if so, cast it to array using CastToArray.
+                        if (obj.GetType() == typeof(ArrayList))
+                        {
+                            return CastUnpickledItems.CastToSimpleArray(unpickledItems);
+                        }
+                        // Break the loop if other types.
+                        else
+                        {
+                            break;
+                        }
+                    }
+                    // Check if array of rows.
+                    if (objArr.GetType() == typeof(RowConstructor))
+                    {
+                        Row row = (objArr as RowConstructor).GetRow();
+                        if (row.Values[0].GetType() == typeof(ArrayList))
+                        {
+                            return CastUnpickledItems.CastToRowArray(unpickledItems);
+                        }
+                        else
+                        {
+                            break;
+                        }
+                    }
+                    else
+                    {
+                        break;
+                    }
```

Comment:
Do we need all these if/else/continue/breaks?  This seems fishy and the code/logic can be redone to avoid all this ?

Reply:
You are right, this part needs to be redone. I will update tonight.

Reply:
Redone.